modules/dnn/src/layers/nll_loss_layer.cpp

```diff
@@ -0,0 +1,223 @@
+            wflat.convertTo(w_f, CV_32F);
+        }
+
+        auto weightAt = [&](int c) -> float {
+            if (!hasWeight) return 1.f;
+            CV_DbgAssert(0 <= c && c < C);
+            return w_f.at<float>(c, 0);
+        };
+
+        const int nstripes = 16;
+        Mat logp32; logp.convertTo(logp32, CV_32F);
+        const size_t sN = logp32.step1(0);
+        const size_t sC = logp32.step1(1);
+
+        CV_Assert(label.depth() == CV_32S || label.depth() == CV_64S);
+        CV_Assert((int)label.total() == N*S);
+        Mat idx1D = label.reshape(1, N*S);
+
+        Mat per(N*S, 1, CV_32F, Scalar(0));
+        Mat validMask(N*S, 1, CV_8U, Scalar(1));
+        Mat effW(N*S, 1, CV_32F, Scalar(1));
+
+        if (label.depth() == CV_32S)
+        {
+            reduceNLLPerSample<int32_t>(logp32, sN, sC, S, idx1D, C, ignoreIndex, weightAt, per, effW, validMask, nstripes);
+        }
+        else
+        {
+            reduceNLLPerSample<int64_t>(logp32, sN, sC, S, idx1D, C, ignoreIndex, weightAt, per, effW, validMask, nstripes);
+        }
+
+        if (!isReduced)
+        {
+            per.reshape(1, out_loss.size).copyTo(out_loss);
+        }
+        else
+        {
+            double num = 0.0, den = 0.0;
+            const float* perData = per.ptr<float>();
+            const float* effData = effW.ptr<float>();
+            const uchar* validData = validMask.ptr<uchar>();
+            const int R = per.rows;
+
+            for (int r = 0; r < R; ++r) {
+                const float m = validData[r] ? 1.f : 0.f;
+                num += static_cast<double>(perData[r] * m);
+
+                if (reduction == DNN_LOSS_REDUCTION_MEAN)
+                {
+                    den += static_cast<double>(std::max(1e-12f, effData[r]) * m);
+                }
+            }
+            const float out = (reduction == DNN_LOSS_REDUCTION_SUM) ? static_cast<float>(num) : static_cast<float>((den > 0.0) ? (num / den) : 0.0);
+            out_loss.at<float>(0) = out;
+        }
+    }
+
+private:
+    template<typename T, typename WeightGetter>
+    static inline void reduceNLLPerSample(const Mat& logp32,
+                                          const size_t sN,
+                                          const size_t sC,
+                                          const int S,
+                                          const Mat& idx1D,
+                                          const int C,
+                                          const int ignoreIndex,
+                                          WeightGetter weightAt,
+                                          Mat& per,
+                                          Mat& effW,
+                                          Mat& validMask,
+                                          const int nstripes)
+    {
+        CV_Assert(idx1D.cols == 1);
+        parallel_for_(Range(0, idx1D.rows), [&](const Range& rr){
+            const float* base = logp32.ptr<float>();
```

Comment:
store values, accessed inside the innermost loop inside parallel_for_:

```
const T* idx1DData = idx1D.ptr<T>();
const float* wfData = w_f.ptr<float>();
uchar* validMaskData = validMask.ptr<uchar>();
float* errWData = errW.ptr<float>();
bool hasWeight = hasWeight_;
const int ignoreIndex = ignoreIndex_;
const in C = C_;
... // also S, sN, sC
```

also, move weightAt implementation inline into the parallel_for_ loop body.